question-answering-using-openai/main.bal

```
@@ -0,0 +1,153 @@
+    }
+
+    string instruction = "Answer the question as truthfully as possible using the provided context, and if the answer is not contained within the text below, say \"I don't know.\"\n\nContext:\n";
+    return string `${instruction} ${context} \n\n Q: ${question} \n A:`;
+}
+
+function generateAnswer(string prompt) returns string|error {
+    openai:CreateCompletionRequest prmt = {
+        prompt: prompt,
+        model: "text-davinci-003"
+    };
+    openai:CreateCompletionResponse completionRes = check openaiClient->/completions.post(prmt);
+    return completionRes.choices[0].text ?: "";
+}
+
+function loadData(string sheetId, string sheetName = "Sheet1") returns [map<string>, map<decimal[]>]|error {
+    // Fetch the data from the 'heading' and 'content' columns.
+    sheets:Range range = check spreadsheetClient->getRange(sheetId, sheetName, "A2:B");
+
+    // Define an empty dictionaries for documents and embeddings.
+    map<string> documents = {};
+    map<decimal[]> doc_embeddings = {};
+
+    // Iterate through the array of arrays and populate the dictionaries with the content and embeddings for each doc.
+    foreach any[] row in range.values {
+        string title = <string>row[0];
+        string content = <string>row[1];
+
+        documents[title] = content;
+        doc_embeddings[title] = check getEmbedding(title + "\n" + content);
+    }
+    return [documents, doc_embeddings];
+}
+
+// Load the data and compute the embeddings when the service starts
+[map<string>, map<decimal[]>] [documents, doc_embeddings] = check loadData(sheetId, sheetName);
+
+service / on new http:Listener(8080) {
+
+    resource function post generateAnswer(@http:Payload Request request) returns Response {
+
+        string question = request.question;
+        string|error prompt = constructPrompt(question, documents, doc_embeddings);
+        if prompt is error {
+            log:printError("Error constructing prompt: ", prompt);
```

Instead of checking and returning empty answer, use `check` to return error directly.

---

addressed in [fa5629f](https://github.com/ballerina-guides/ai-samples/pull/2/commits/fa5629f9ca6bfae66689af4f34151a7883b405d2)